### PR TITLE
Added error handling to setup.

### LINKED ISSUE #355: Setup : Handle exceptions and display user friendly error message
Handle the exceptions that can happen during the setup.

Example: If a wrong connection string is passed or if the database doesn't exist, the setup page shoud display an error message instead of displaying the error page.

### PR DESCRIPTION
**Pull request by sfmskywalker**
This fixes an issue where Orchard crashes when it is unable to connect to the database (e.g. due to incorrect connection string).
Whenever an exception occurs during DB initialization, that error is shown on the setup screen.

Fixes #355 

image

### CODE REVIEW

File: src/Orchard.Cms.Web/Modules/Orchard.Setup/Services/SetupService.cs [hunk 1]
```diff
@@ -135,14 +135,17 @@ ILogger<SetupService> logger
...
                         // Tables already exist or database was not found
 
                         // The issue is that the user creation needs the tables to be present,
                         // if the user information is not valid, the next POST will try to recreate the
-                        // tables. The tables should be rollbacked if one of the steps is invalid,
+                        // tables. The tables should be rolled back if one of the steps is invalid,
                         // unless the recipe is executing?
+
+                        context.Errors.Add("DatabaseProvider", $"An error occurred while initializing the datastore: {ex.Message}");
+                        return null;
```
Comment (agriffard) [inline src/Orchard.Cms.Web/Modules/Orchard.Setup/Services/SetupService.cs]: @Jetski5822 doesn't like this return null;

File: src/Orchard.Cms.Web/Modules/Orchard.Setup/Views/Setup/Index.cshtml [hunk 2]
```diff
@@ -106,6 +107,7 @@
                 <option value="@provider.Value" data-connection-string="@provider.HasConnectionString">@provider.Name</option>
                 }
             </select>
+            <span asp-validation-for="DatabaseProvider" class="text-danger"></span>
```
Comment (agriffard) [inline src/Orchard.Cms.Web/Modules/Orchard.Setup/Views/Setup/Index.cshtml]: May be we can find a way to have only one <span asp-validation-for="DatabaseProvider" class="text-danger"></span>

**Comment by agriffard** [comment]
Ok. It now handles error during database configuration step.

Here is what it displays on my machine if the database doesn't exist :

image

The message is longer.

Just a question: Do we have to display the sql exception message to the user?

**Comment by agriffard** [comment]
There is also a specific setup screen for the tenant in Orchard.Tenants\Views\Create.cshtml and Edit.cshtml

I guess it needs the same fix.

**Comment by agriffard** [comment]
BTW, It is good to see you working an Orchard Core ;-)

File: src/Orchard.Cms.Web/Modules/Orchard.Setup/Services/SetupService.cs [hunk 3]
```diff
@@ -135,14 +135,17 @@ ILogger<SetupService> logger
...
                     {
                         // Tables already exist or database was not found
 
                         // The issue is that the user creation needs the tables to be present,
                         // if the user information is not valid, the next POST will try to recreate the
-                        // tables. The tables should be rollbacked if one of the steps is invalid,
+                        // tables. The tables should be rolled back if one of the steps is invalid,
                         // unless the recipe is executing?
+
+                        context.Errors.Add("DatabaseProvider", $"An error occurred while initializing the datastore: {ex.Message}");
```
Comment (agriffard) [inline src/Orchard.Cms.Web/Modules/Orchard.Setup/Services/SetupService.cs]: Please injectIHtmlLocalizer and add `T[ ]` around.

We don't use a lot string interpolation but at some point, we will have to be sure that it works with T[] localizer

File: src/Orchard.Cms.Web/Modules/Orchard.Setup/Services/SetupService.cs (hunk 1 above)
Comment (sfmskywalker) [inline src/Orchard.Cms.Web/Modules/Orchard.Setup/Services/SetupService.cs]: Would he rather return an empty string instead? Little point in generating an execution ID at this point when the DB connection fails, right?

Comment (sfmskywalker) [inline src/Orchard.Cms.Web/Modules/Orchard.Setup/Services/SetupService.cs]: Also, screw nick. ;)

File: src/Orchard.Cms.Web/Modules/Orchard.Setup/Services/SetupService.cs (hunk 3 above)
Comment (sfmskywalker) [inline src/Orchard.Cms.Web/Modules/Orchard.Setup/Services/SetupService.cs]: I should do that, thanks.

File: src/Orchard.Cms.Web/Modules/Orchard.Setup/Views/Setup/Index.cshtml (hunk 2 above)
Comment (sfmskywalker) [inline src/Orchard.Cms.Web/Modules/Orchard.Setup/Views/Setup/Index.cshtml]: We probably can, however I'm not sure. Ideally, the error message would span the full col-12, but if I add a wrapping `div`, that div will obviously be rendered regardless of there being an error or not. Any suggestions?

**Comment by sfmskywalker** [comment]
> Just a question: Do we have to display the sql exception message to the user?

On the one hand, I think it's very helpful to see what's going on. On the other hand, that's what we have log  files for, so we could also simply display a "Error connecting to DB, check error log for details" kind of message instead. Maybe that's better. What do you think?

**Comment by sfmskywalker** [comment]
> There is also a specific setup screen for the tenant in Orchard.Tenants\Views\Create.cshtml and Edit.cshtml

Good catch. Added the error message span there as well.

**Comment by sfmskywalker** [comment]
> There is also a specific setup screen for the tenant in Orchard.Tenants\Views\Create.cshtml and Edit.cshtml

I reverted my changes for the tenant views, because the DB connection string isn't actually used on those screens; only when running setup on a tenant, will the connection string be used; which in turn does use the DB connection string.

**Comment by sebastienros** [comment]
I would have displayed the error message directly, not using the log. Because, this is exactly what you want to do when you can't connect to your database, know why, and not have to understand what and where the log is.

**Comment by sfmskywalker** [comment]
I prefer that as well. Let me send another PR.

**Pull request merged**